Feature/composite src generator

badcel opened the pull request:
WIP

badcel commented:
@firox263  @na2axl 

This is the follow up PR for #133 

It adds a source generator and modifies the sample project which demonstrates the usage of a composite template.

What do you think?

badcel commented:
Depends on #140

badcel commented:
Hm as we said that we wan't the source generators to be opt in, I think we should have two samples. One using a source generator for composite templates and one without. What do you guys think @gircore/coreteam ?

firox263 commented:
I like that idea. It's good understanding how everything works under the hood without any "magic", even if source generators are the practical option that most people would use.

I imagine there are some edge cases where doing it manually could be helpful too.

badcel commented:
I wonder if we can have a better name for the `Integrator` project? As it would be some public nuget package later on. Any ideas?

badcel commented:
We probably only want to provide one source generator to make it easy for users of our library to get the C# coding benefits. So the name should probably be s.th. general but with a connection to us.

As our libraries have a package id like `Gir.CS.X` it could be `Gir.CS.Integration`. Actually I like `Integration` better than `Integrator`. So I would rename the project accordingly?

@firox263 thoughts?

firox263 commented:
I like `Integration` quite a bit. I think we should go with that 👍

Off-topic, but what's the reason we have `CS` in our package IDs? It seems a bit redundant.

badcel commented:
@firox263 / @na2axl :+1: 

I renamed the folder and added some common `*.props` files to avoid duplication.

Regarding the package id: It was meant as preparation for libraries of a different language. Eg. if we manage to build some F# bindings on top of our C# ones to improve the language integration, those packages could be `Gir.FS.X`.

What do you guys think of it? Should we keep it or should we use a totally different name?

firox263 commented:
> Regarding the package id: It was meant as preparation for libraries of a different language. Eg. if we manage to build some F# bindings on top of our C# ones to improve the language integration, those packages could be Gir.FS.X.

A while back, I decided to see if I could port QuickStart to visual basic, mostly to see if it was possible. It actually worked really well. Turns out, as long as we are CLR compliant, we can use C# libraries from other .NET languages "for free". So I'm not sure we'd ever need to build dedicated F# bindings.

One option could be to make a `Gir.Integration.CSharp`, and a `Gir.Integration.FSharp`, etc and use that as a namespace for all our language-dependent integrations. I haven't looked into this PR in detail enough to know if it works like that, but it could be an option for code sharing. What do you think?

badcel commented:
> A while back, I decided to see if I could port QuickStart to visual basic, mostly to see if it was possible. It actually worked really well. Turns out, as long as we are CLR compliant, we can use C# libraries from other .NET languages "for free".

Yep regarding VB there is no separation needed. The same goes for F# in principle. But I could guess that a nice F# API looks different to the C# one as we could try to hide objects and make it _somehow_ functional (perhapps by wrapping the C# API).

I never tried it or thought it completely through but regarding functional UIs there are some [slides](https://de.slideshare.net/frankakrueger/functional-guis) from @praeclarum.

> One option could be to make a Gir.Integration.CSharp, and a Gir.Integration.FSharp, etc and use that as a namespace for all our language-dependent integrations.

This sounds like a good idea. I'm going to update this PR with your idea.

> Alternatively, if we wanted to keep our current naming, I'd prefer to put the language after the library (e.g. Gir.Gtk.CSharp, Gir.Gtk.FSharp) so we can sort by library alphabetically.

As a nuget search for `CSharp` brought a lot of hits and the `CSharp` was often a suffix and not a prefix I would do like you suggested and make it a suffix.

firox263 reviewed (approved):
Brilliant work. Very little issues here, only some small stylistic ones.

Could you add the new samples to the solution before you merge?

Also, both samples have the same name of `CompositeTemplates.csproj`, which makes it hard to browse/run/debug with VS. We should probably call them `UsingSourceGenerator.csproj` and `NoSourceGenerator.csproj` respectively.

badcel commented on Tests/Build.Tests/RunnerTest.cs:
```diff
@@ -83,19 +105,21 @@ public void InvokingBuildTargetExecutesLibraryGeneratorAndLibraryBuilder(string
 
         [TestMethod]
-        public void InvokingSamplesTargetExecutesLibraryBuilderAndSampleBuilder()
+        public void InvokingSamplesTargetExecutesLibraryBuilderIntegrationBuilderSampleBuilder()
```
I think I leave it like it is for now. But I agree that we can try to find better names. I would suggest working on this in #180 and fix it if we find some kind of convention which we want to use.

badcel commented on Libs/GObject/Extensions/AssemblyExtension.cs:
```diff
@@ -7,7 +7,7 @@ namespace GObject
     public static class AssemblyExtension
     {
-        public static Bytes ReadResource(this Assembly assembly, string resource)
+        public static Bytes ReadResourceBytes(this Assembly assembly, string resource)
```
Done :+1:

badcel commented on Libs/Gtk3/Classes/TemplateAttribute.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    using System;
+
+    [AttributeUsage(AttributeTargets.Class)]
```
Done :+1:

badcel commented:
@firox263 Thanks for the review :+1:
